### DISCUSSION TITLE
Is there a way to use an external file as either the complete or partial body in a request?

### DISCUSSION THREAD

**Discussion by nerdo** [mistweaverco/kulala.nvim#72] [Q&A] [answered]
If this is already a thing, I wasn't able to find documentation on it, so any help is appreciated.

If not, can this be turned into a feature request?

I think it would be incredibly useful to refer to an external file and use it as the entire or partial request body. Some apis require either large payloads or some just have parts that are repetitive and could be re-used among one or more request bodies.

It would be nice to define it in one place and reference it in the request body somehow.

**Comment by gorillamoe** [comment]
Hey there, thanks for coming up with this. There's indeed missing documentation on this.

Should be already possible like this:

```http
POST https://httpbin.org/post HTTP/1.1
content-type: application/json
accept: application/json

< ./some.json
```

Does this work for you and suffice?

**Comment by gorillamoe** [comment]
@nerdo, does this work for you?

**Comment by nerdo** [comment]
I will try this today. Is it possible to do partial json? i guess i'll try
that to but wondering what an example of that syntax might be.

**Comment by nerdo** [comment]
Ok, so I just tried it with a partial payload and it did not seem to work.

Here's what I tried. This is a real-world example using the sendgrid api where I would like to test sending to different recipients using variables, and want the template data to be pulled in from json so that I can easily test different templates that require different data without too much fuss.

```
@templateId=something
@from=bob@example.com
@to=someone@example.com

POST https://api.sendgrid.com/v3/mail/send
Authorization: Bearer {{API_KEY}}
Content-type: application/json

{
	"from": {
		"email": "{{from}}"
	},
	"personalizations": [
		{
			"to": [
				{"email": "{{to}}"}
			],
			"dynamic_template_data": < ./template-data/go-live.json
		}
	],
	"template_id": "{{templateId}}"
}

```

BTW I tested this inline (i.e. without the `< ./template-data/go-live.json` but with the data in the body) and it was working, but I am getting an error response from sendgrid.

On that note, it is great that we're able to see the response headers with this plugin but it would also be helpful to view the full request that goes over the wire (or even just the curl command that ran it or something of that nature) to help debug some of the more dynamic aspects of the request building.

I suspect that the `< ./template-data/go-live.json` isn't getting parsed, but with the ability to see the request, I'd be able to confirm that.

**Reply by gorillamoe** [reply]
Inline isn't working ATM. You can see the request with `:lua require("kulala").copy()` and paste it somewhere.

But as I said in lining is not (yet) supported.

The parser expects the `<` to be the first character of a body line.

So maybe this here might work

```json
{
	"from": {
		"email": "{{from}}"
	},
	"personalizations": [
		{
			"to": [
				{"email": "{{to}}"}
			],
< ./template-data/go-live.json
		}
	],
	"template_id": "{{templateId}}"
}

```

**Reply by gorillamoe** [reply, marked as answer]
This `<` at the start of the line conforms to the http file spec afaik. But it's not very readable in this case.

I might add another metadata field for that, something like `# @file-to-env name_of_env ./filename.json`

Which then could be used as a normal env `{{name_of_env}}`..

Not yet implemented, but might work quite well.

**Reply by nerdo** [reply]
Ah, that does work when I put it at the beginning of the line. Thanks!

As you mentioned, it's not as readable - file to env or even just a way to specify it using a variable (not sure if env and variables are considered the same thing conceptually).

Thanks again!

**Reply by gorillamoe** [reply]
Implemented in https://github.com/mistweaverco/kulala.nvim/releases/tag/v2.6.0
